compatibility/compatibility-tests/src/test/java/org/projectnessie/tools/compatibility/tests/ITRollingUpgrades.java

```diff
@@ -0,0 +1,209 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+import com.google.common.collect.Maps;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.projectnessie.client.api.NessieApiV1;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.CommitMeta;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.Operation.Put;
+import org.projectnessie.model.Reference;
+import org.projectnessie.tools.compatibility.api.NessieAPI;
+import org.projectnessie.tools.compatibility.api.NessieVersion;
+import org.projectnessie.tools.compatibility.api.TargetVersion;
+import org.projectnessie.tools.compatibility.api.Version;
+import org.projectnessie.tools.compatibility.api.VersionCondition;
+import org.projectnessie.tools.compatibility.internal.RollingUpgradesExtension;
+
+@ExtendWith(RollingUpgradesExtension.class)
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+@VersionCondition(minVersion = "0.31.0")
+public class ITRollingUpgrades {
+
+  public static final String NO_ANCESTOR =
+      "2e1cfa82b035c26cbbbdae632cea070514eb8b773f616aaeaf668e2f0be8f10d";
+  /** Nessie source version of the rolling upgrade. */
+  @NessieVersion Version version;
+
+  /** API of the "old" Nessie version instance. */
+  @NessieAPI(targetVersion = TargetVersion.TESTED)
+  NessieApiV1 api;
+
+  /** API of the "current" (in-tree) Nessie version instance. */
+  @NessieAPI(targetVersion = TargetVersion.CURRENT)
+  NessieApiV1 apiCurrent;
+
+  static List<Reference> createdReferences = new ArrayList<>();
+  static Map<ContentKey, IcebergTable> createdContent = new LinkedHashMap<>();
+
+  @BeforeAll
+  static void cleared() {
+    // The Nessie repository is reinitialized for every rolling-upgrade version combination.
+    createdReferences.clear();
+    createdContent.clear();
+  }
+
+  @Test
+  @Order(100)
+  void defaultBranchInOld() throws Exception {
+    Branch defaultBranch = api.getDefaultBranch();
+    assertThat(defaultBranch).extracting(Branch::getName).isEqualTo("main");
+  }
+
+  @Test
+  @Order(101)
+  void defaultBranchInNew() throws Exception {
+    Branch defaultBranch = apiCurrent.getDefaultBranch();
+    assertThat(defaultBranch).extracting(Branch::getName).isEqualTo("main");
+  }
```

Comment:
It feels like we could probably parameterize these test by `TargetVersion` and maybe use a `ParameterResolver` to inject the right API impl based on the enum value. WDYT?

Reply:
Using `@NessieAPI` as a paramter already works. These tests (and the following) should work in a defined order though.